### DISCUSSION TITLE
assignment 7th 내용 질문 - (Before TODO)

### DISCUSSION THREAD

**Discussion by Lcat12345** [MyeongWoonJang/cpp-study#83] [Q&A] [answered]
> 다음의 경우에 `dynamic allocation`이 유용할 수 있습니다.
- 필요한 메모리의 크기가 큰 경우
  
  - `stack` 메모리는 연속된 메모리 배치를 요구하고 `static`하게 `allocate`되는 만큼,   
  작은 크기로 설정하는 게 일반적입니다.
  - `heap` 메모리는 메모리 배치가 연속적이지 않아도 괜찮고,   
  필요할 때 `allocate`하고 필요없으면 `deallocate`하는 `free memory`입니다.
---

위 글에서 필요한 메모리의 크기가 큰 경우에서 stack은 동적할당이 유용하지 않은 걸까요? 
> 작은 크기로 설정하는게 일반적입니다.

이 설명을 저는 메모리의 크기를 작게 설정하는게 일반적이다로 받아들였습니다.

연속적이다와 비연속적이다는 메모리의 크기가 큰 경우와 어떤 연관이 있는건가요?

---
>`pointer`가 `scope`를 벗어난다 해도, `pointer`가 저장하던 `memory address` 값을 잃는 것이지,   
그 `memory address`에 `allocate` 되었는지 안 되었는지도 모르는 실제 메모리가 `deallocate`되진 않습니다.

이 부분 정말 많이 고민했네요(그래봤자 10분인가 어라)
``` cpp
{
	int* p;
	{
		int a = 3;
		p = &a;
	}
	
	std::cout << p;
}
```
위의 코드에서 범위를 두개를 잡아서(바깥과 안쪽) 제가 이해를 어떻게 했는지 말씀드리겠습니다.

>`pointer`가 저장하던 `memory address` 값을 잃는 것이지

a의 주소값이 범위를 벗어나 없어진다. 

>`memory address`에 `allocate` 되었는지 안 되었는지도 모르는 실제 메모리

실제 메모리.. 실제 메모리에서 많이 이해가 안가네요

a라는 변수의 공간을 의미하는건가요?

**Comment by Woon-2** [comment, marked as answer]
## stack의 크기

- `heap`이 `random`한 메모리를 얻어오는 것에 반해, `stack`은 연속된 메모리를 얻어옵니다.   
과제에서 직접 `stack`의 메모리가 연속되어 있는 것을 확인하셨을 겁니다.

- `stack` 는 `data` 메모리처럼 `static`하게 `allocate`됩니다.   
  `function call`에 의해 `push&pop`되는 `stack frame`은 메모리의 운영을 표현하지, `allocation&deallocation`을 표현하지 않습니다.

  - `compile` 시점에 어떤 `function`이 또 어떤 `function`을 `call`하는지 다 파악이 되기 때문에,   
  `stack`에 필요한 최대 메모리 크기가 어느 정도인지 `compile` 시점에 알 수 있습니다.
  - 그래서 미리 파악된, `stack`에 필요한 최대 메모리 크기를 처음에 다 확보하고,   
  시간 비용이 매우 큰, `system call`을 통한 `allocation&deallocation`을 아예 하지 않습니다.

이제 문제가 눈에 보였을 거라 생각합니다.   
2MB 정도의 메모리를 잡아먹는 `object`가 100개 있으면, 200MB의 연속된 메모리가 필요하게 되죠.   
`OS`는 정해진 메모리를 여러 `process`에 분할해 사용하는데, 최대한 빈 메모리가 없도록 `process`에 필요한 메모리 영역들을 끼워맞춥니다.   

그런데 한 `process`가 자신이 `run`되는 내내 무지막지하게 큰 메모리 공간을 잡아먹고 있다고 생각해보세요.    

`heap`에서 2MB 정도의 메모리를 잡아먹는 `object`를 100개 생성하면,   
메모리를 200MB의 덩어리로 두는 것 외에 몇 조각으로 나눠서 다른 메모리 영역에 끼워 맞추거나,   
틈을 두어서 다른 `process`를 위한 메모리 영역을 구비해둘 수 있습니다.

필요할 때 `allocate`했다가 필요 없어지면 `deallocate`하면 되니, 언제까지고 메모리를 잡아먹고 있을 메모리 괴물도 아니죠.

그래서 필요한 메모리 크기가 클 경우에 `stack` 대신 `heap`을 이용하는 겁니다.

## pointer가 저장하는 memory address의 실제 메모리

글에 택배원 비유를 해놨었는데 확 와닿진 않았나 보네요.

`A is a pointer to B`라고 할 때,
`A`의 소멸과 `B`의 소멸이 독립적이라는 설명입니다.

`A`가 소멸해도 `B`에는 영향이 없고, `B`가 소멸해도 `A`에는 영향이 없습니다.

<br>

실제 메모리라는 표현을 한 것은 이런 연유에서입니다.

어떤 `object`의 `memory address`를 출력하면 `0xabcdef`가 나오죠.   
이는 단지 `object`가 차지하는 메모리의 시작 번지에 불과하지,   
`object`가 얼만큼의 메모리를 차지하는지와 그 메모리를 어떻게 이용하는지와는 상관이 없습니다.

거기에 메모리를 `allocate`했는지, 했으면 얼만큼 했는지, `0xabcdef`~`0xabcef0`을 차지하고 있는지,   
그 메모리가 `int`를 위한 메모리인지 `float`를 위한 메모리인지, 이런 건 전부 `memory address` 그 자체와는 무관한 정보라는 뜻입니다.

PC방의 컴퓨터를 한번 생각해봅시다.

1번 컴퓨터, 2번 컴퓨터, 3번 컴퓨터, ... 이렇게 컴퓨터가 많겠죠.

1번 컴퓨터에 오늘 아침에는 `Bjarne Stroustrup`이 앉아 있었고,   
오늘 점심에는 `Dennis MacAlistair Ritchie`가,   
오늘 저녁에는 아무도 앉아있지 않았습니다.

`Bjarne Stroustrup`이 이용 시간이 끝났다고 해서 1번 컴퓨터를 들고 가버리나요?   
아니죠, 이제 그 자리에는 또 컴퓨터가 필요한 사람이 찾아와서 앉습니다.   
그래서 `Dennis MacAlistair Ritchie`가 앉았죠.

`memory address`는 1번 컴퓨터, 실제 메모리는 거기에 앉는 사람들인 겁니다.
여러분이 `OS` 즉, PC방 사장이라면 1번 컴퓨터에 사람을 앉히는 일에만 관심이 있지   
그 사람이 누구인지는 관심이 없는 것입니다.

또한 어느날 갑자기 PC방이 망한다 해도   
1번 컴퓨터가 사라지는 것이지, `Bjarne Stroustrup`과 `Dennis MacAlistair Ritchie`가 싸늘한 주검이 되는 것이 아닙니다.   
그들은 다른 PC방을 찾아가겠죠.

이 내용이   
>`pointer`가 `scope`를 벗어난다 해도, `pointer`가 저장하던 `memory address` 값을 잃는 것이지,
>그 `memory address`에 `allocate` 되었는지 안 되었는지도 모르는 실제 메모리가 `deallocate`되진 않습니다.

**Comment by Lcat12345** [comment]
Stack은 연속적인 메모리라 부담인거군요

실제 메모리에 대한 내용도 이해됬습니다

답변 감사합니다

**Reply by Woon-2** [reply]
`Mark as answer` 버튼 눌러주세용.   
`stack` 메모리가 연속적인 메모리인 점과 `stack` 메모리가 `static`하다는 점 중에서는   
`stack` 메모리가 `static`하다는 점이 훨씬 부담이 됩니다.   

워낙 잘 학습하시니 이해에 큰 어려움은 없었을 거라 생각되지만,
혹여 완벽한 이해가 힘드셨다 하면 말씀해주세요.   
잘못 알거나 대충 알고 가는 게 모르고 가는 것보다 악수가 됩니다.
